// @onscroll/di-js/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _injector = require('./injector');

Object.defineProperty(exports, 'Injector', {
  enumerable: true,
  get: function get() {
    return _injector.Injector;
  }
});

var _annotations = require('./annotations');

Object.defineProperty(exports, 'annotate', {
  enumerable: true,
  get: function get() {
    return _annotations.annotate;
  }
});
Object.defineProperty(exports, 'Inject', {
  enumerable: true,
  get: function get() {
    return _annotations.Inject;
  }
});
Object.defineProperty(exports, 'InjectLazy', {
  enumerable: true,
  get: function get() {
    return _annotations.InjectLazy;
  }
});
Object.defineProperty(exports, 'InjectPromise', {
  enumerable: true,
  get: function get() {
    return _annotations.InjectPromise;
  }
});
Object.defineProperty(exports, 'Provide', {
  enumerable: true,
  get: function get() {
    return _annotations.Provide;
  }
});
Object.defineProperty(exports, 'ProvidePromise', {
  enumerable: true,
  get: function get() {
    return _annotations.ProvidePromise;
  }
});
Object.defineProperty(exports, 'SuperConstructor', {
  enumerable: true,
  get: function get() {
    return _annotations.SuperConstructor;
  }
});
Object.defineProperty(exports, 'TransientScope', {
  enumerable: true,
  get: function get() {
    return _annotations.TransientScope;
  }
});
Object.defineProperty(exports, 'ClassProvider', {
  enumerable: true,
  get: function get() {
    return _annotations.ClassProvider;
  }
});
Object.defineProperty(exports, 'FactoryProvider', {
  enumerable: true,
  get: function get() {
    return _annotations.FactoryProvider;
  }
});
