// @onscroll/logger-ts/src/TransportFactory/Transports/StatsDTransport/StatsDTransport.ts
import {Inject} from "@onscroll/di-ts";
import {AbstractTransport} from "../AbstractTransport";
import {StatsDClientFactory} from "./StatsDClientFactory/StatsDClientFactory";


@Inject
export class StatsDTransport extends AbstractTransport {

    private statsDConnector: any;


    constructor(private statsDConnectorFactory: StatsDClientFactory) {

        super();

    }


    public initialize(settings) {

        super.initialize(settings);
        this.statsDConnector = this.statsDConnectorFactory.createClient(settings);

    }


    public increment(fieldName: string, delta: number) {

        this.statsDConnector.increment(this.getFieldName(fieldName), delta);

    }


    public decrement(fieldName: string, delta: number) {

        this.statsDConnector.decrement(this.getFieldName(fieldName), delta);

    }


    public gauge(fieldName: string, value: number) {

        this.statsDConnector.gauge(this.getFieldName(fieldName), value);

    }


    public gaugeDelta(fieldName: string, value: number) {

        this.statsDConnector.gaugeDelta(this.getFieldName(fieldName), value);

    }


    private getFieldName(fieldName: string) {

        return this.settings.name + "." + fieldName;

    }
}
